calnex/cert/cert.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright (c) Meta Platforms, Inc. and affiliates.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cert
+
+import (
+	"bytes"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// Bundle holds details of a key + its certificates
+type Bundle struct {
+	Certs   []*x509.Certificate
+	PrivKey *rsa.PrivateKey
+}
+
+// Fetch connects to the remote host specification (hostname + TCP port) and retrieves
+// the remote TLS certificate, returning it as a Bundle.
+func Fetch(host string) (*Bundle, error) {
+	conf := &tls.Config{
+		InsecureSkipVerify: true,
+	}
+
+	conn, err := tls.Dial("tcp", host, conf)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	bundle := &Bundle{
+		Certs: conn.ConnectionState().PeerCertificates,
+	}
+
+	return bundle, nil
+}
+
+// Parse parses the supplied byte array to check it is a valid x509 PEM file.
+// If this is successful the certificate bundle (key + any certs) is returned.
+// If not an error is returned.
+func Parse(data []byte) (*Bundle, error) {
+	bundle := &Bundle{}
+
+	for len(data) > 0 {
+		var block *pem.Block
+
+		block, data = pem.Decode(data)
+		if block == nil {
+			return nil, fmt.Errorf("failed to parse certificate PEM")
+		}
+
+		if block.Type == "CERTIFICATE" {
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, err
+			}
+			bundle.Certs = append(bundle.Certs, cert)
+		} else if strings.Contains(block.Type, "PRIVATE KEY") {
+			if bundle.PrivKey != nil {
+				return nil, fmt.Errorf("multiple private keys in PEM")
```

Comment:
Please define errors and in unit tests check for the exact errors returned the invalid key.